### DISCUSSION TITLE
Deploying the app

### DISCUSSION THREAD

**Discussion by MenJeremy** [nickcoutsos/keymap-editor#22] [General]
Hi,

I am wondering if it's possible to have some instructions on how to deploy the app, can you do it with GitHub Page ?

Do you need to host the NodeJS separately, if so what would be some steps, sorry I am still new to NodeJS :)

Thanks for your help !!

Jeremy

**Comment by nickcoutsos** [comment]
There's a few different parts, it might be helpful if you could explain how you're using the editor currently and what you want to do (ie, integration with GitHub repos vs local file writing). There is:

* the frontend application, which can be deployed via GitHub Pages
* the backend API, a NodeJS app (which I deployed to Heroku)
* the GitHub App, which is the integration that allows this API to access GitHub repositories that have enabled it

At this time the GitHub integration is the main supported use case and there's a little bit of documentation here https://github.com/nickcoutsos/keymap-editor/tree/master/api/services/github for getting that set up. Not complete instructions by any means, but it gives a little overview about the integration.

**Reply by MenJeremy** [reply]
Thanks, I appreciate the reply!

What I would like to do is to replicate the project as is (with GitHub integration), but adding macro functionality.
So far I was able to run it locally (on a Windows machine), load macros (read-only) and even integrate with a GitHub App.
Next step I would like to do is figure out how to setup the deployment.

The thing that I am still trying to wrap my head around is how to separate the frontend and backend when deploying.
If I want to deploy the backend, which part of the project do I have to build ?
If I want to deploy the frontend, which part needs to be build ?

Thanks in advance !

**Reply by nickcoutsos** [reply]
For the frontend you want to go into the `application/` directory and build that. If you look at `application/config.js` it references a number of environment variables you'll need to set during the build (I guess how to do that with windows depends on your terminal app, I always forget).

The backend is pretty much already set up for heroku. If you create an app there and set the appropriate environment variables (see the document I linked previously) then heroku will take care of the build step when deploying.

Also, if you're interested in contributing here I'm open to reviewing a pull request.

**Reply by MenJeremy** [reply]
Thank you! Will test it out, appreciate your help!

**Reply by nickcoutsos** [reply]
Hi @MenJeremy 

I see you've done some work on macro editing and have it merged into KinesisCorporation's fork, I wonder if I might be able to discuss collaboration on the keymap editor to combine efforts and increase the benefit for more users. I have added quite a bit since the forks started, particularly in making inline changes to the keymap file and everything that unlocks, and I'd hate to see anyone's work go to waste when a lot of these use cases can be supported modularly.

Let me know if you'd like to talk!

**Comment by MenJeremy** [comment]
Hello Nick,

You should have my direct email in my reply.

I've been working for Kinesis for many years now, and when they asked me to
work on the web app I accepted but had to learn NodeJs and Vue.js, it's
been a big learning curve as they wanted to deliver the product quickly.
It's also my first time working from an existing github repository, so not
sure if I'm doing all of this correctly, you might be able to give me some
pointers there.

I saw that you've updated your repo quite a bit and switched from Vue to
React.
Are you suggesting that I could fork from your updated repository and
integrate macros in there ?
I don't mind doing this, but it all depends on how much effort that would
involve.

Let me know what you propose and we can go from there.

Oh and I have an extra question for you.
I noticed that the list of repositories that are proposed come from the
hoster (i.e. Kinesis), so if any user makes changes, they are not saved in
their own repositories but in the hoster config repository, is that normal ?

Thanks,

Jeremy

**Reply by nickcoutsos** [reply]
Hey Jeremy,

Nice to hear back :)

GitHub doesn't give me your email address but if you'd prefer to speak more directly you're free to contact me at ncoutsos@gmail.com

The short version is that if people want to expand the functionality of this tool I would love for it to happen in a way that's coordinated and maximizes benefit to all users. You'll want to support your customers, certainly, but there's no reason this can't support anybody with a ZMK-powered keyboard.

Our paths have drifted quite bit, so merging things won't be trivial (there's some review I'd want to do as well) but it's doable. I understand that your time is limited and you likely have other priorities so we'd have to figure that out together. It may help to talk about goals/vision for the editor? From the looks of things, macros are pretty important to the product? I've never used them myself but I've been trying to work towards combo editing so hopefully that's two useful features in the bag :)

> Oh and I have an extra question for you.
> I noticed that the list of repositories that are proposed come from the
> hoster (i.e. Kinesis), so if any user makes changes, they are not saved in
> their own repositories but in the hoster config repository, is that normal ?

I'm not sure I'm following... the typical series of events is

1. User authenticates with GitHub
2. User authorizes the keymap editor GitHub App to connect to their account
3. User selects which repositories they want the app to access
4. (redirect back to keymap editor)
5. Keymap Editor API fetches user's repos and keyboard info from GitHub API
6. User makes changes and submits
7. Keymap Editor API commits changes to user's repo via GitHub API

I haven't looked at your changes in a lot of detail but I don't expect that has changed. The important detail, though, is that the user wouldn't even have write permissions to commit changes to the hoster repository (except perhaps in your case where you're part of the KinesisCorporation organization). Sorry, I'm not sure I'm answering your question as I'm a little unclear what's being asked.d
